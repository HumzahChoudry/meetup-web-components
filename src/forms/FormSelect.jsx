import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

const FormSelect = props => {
	const {
		id,
		name,
		label,
		labelClassName,
		className,
		options,
		disabled,
		selectedOption,
		onChange,
		required,
		error,
		...other
	} = props;

	const classNames = cx(
		{ 'field--error': error },
		className
	);

	const labelClassNames = cx(
		'label--field',
		{ required, disabled },
		labelClassName
	);
	const renderedOptions = options && options.length ? options.map((option, index) => {
		const attrs = {
			key: index,
			selected: option === selectedOption,
			value: option
		};
		return (<option {...attrs}>{option}</option>);
	}) : null;

	return (
		<div>
			{label &&
				<label className={labelClassNames} htmlFor={id}>
					{label}
				</label>
			}
			<select
				name={name}
				required={required}
				className={classNames}
				onChange={onChange}
				disabled={disabled}
				id={id}
				{...other}
			>
				{renderedOptions}
			</select>
			{ error && <p className='text--error'>{error}</p> }
		</div>
	);
};

FormSelect.propTypes = {
	name: PropTypes.string,
	label: PropTypes.string,
	labelClassName: PropTypes.string,
	className: PropTypes.string,
	options: PropTypes.arrayOf(String).isRequired,
	disabled: PropTypes.bool,
	selectedOption: PropTypes.string,
};

export default FormSelect;
