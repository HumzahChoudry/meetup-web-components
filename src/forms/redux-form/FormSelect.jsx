import React from 'react';
import FormSelect from '../FormSelect';

/**
 * This component wraps the standard FormSelect component for use with redux-form.
 * It deconstructs props that redux-form sets and re-sets them on FormSelect.
 * @param {object} props React component props
 * @return {React.Component} FormSelect
 */
const ReduxFormSelect = props => {
	const { meta, validateAfterTouched, ...other } = props;

	const error = (!validateAfterTouched || meta.touched) ? meta.error : null;

	return <FormSelect error={error} {...other} />;
};

ReduxFormSelect.displayName = 'ReduxFormSelect';

export default ReduxFormSelect;

