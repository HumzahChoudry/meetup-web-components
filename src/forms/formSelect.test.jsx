import React from 'react';
import { shallow } from 'enzyme';
import FormSelect from './FormSelect';

const onChangeMock = jest.fn();
const FAVORITE_LINE = 'If thou dost seek to have what thou dost hide,';
const MOCK_OPTIONS = [
	'Love is my sin and thy dear virtue hate',
	'Hate of my sin, grounded on sinful loving:',
	'O, but with mine compare thou thine own state,',
	'And thou shalt find it merits not reproving;',
	'Or, if it do, not from those lips of thine,',
	'That have profaned their scarlet ornaments',
	'And seal\'d false bonds of love as oft as mine,',
	'Robb\'d others\' beds\' revenues of their rents.',
	'Be it lawful I love thee, as thou lovest those',
	'Whom thine eyes woo as mine importune thee:',
	'Root pity in thy heart, that when it grows',
	'Thy pity may deserve to pitied be.',
	FAVORITE_LINE,
	'By self-example mayst thou be denied!'
];
const MOCK_PROPS = {
	id: 'sonnet',
	name: 'sonnets',
	label: 'Shakespearean Sonnet',
	labelClassName: 'labelForSonnet',
	className: 'classForSonnet',
	options: MOCK_OPTIONS,
	onChange: onChangeMock,
	selectedOption: FAVORITE_LINE,
	disabled: true,
	required: true
};
describe('FormSelect', () => {
	it('renders correctly', () => {
		expect(shallow(<FormSelect {...MOCK_PROPS} />)).toMatchSnapshot();
	});
	describe('passing props', () => {
		const component = shallow(<FormSelect {...MOCK_PROPS} />);
		const select = component.find('select');
		it('should have the correct id', () => {
			expect(select.prop('id')).toEqual(MOCK_PROPS.id);
		});
		it('should have the correct name', () => {
			expect(select.prop('name')).toEqual(MOCK_PROPS.name);
		});
		it('should have the correct required status', () => {
			expect(select.prop('required')).toEqual(MOCK_PROPS.required);
		});
		it('should have the correct disabled status', () => {
			expect(select.prop('disabled')).toEqual(MOCK_PROPS.disabled);
		});
		it('should have the correct label', () => {
			expect(component.find('label').text()).toEqual(MOCK_PROPS.label);
		});
	});
	describe('onChange', () => {
		it('calls onChange when option is selected', () => {
			const select = shallow(<FormSelect {...MOCK_PROPS} />).find('select');
			expect(onChangeMock).not.toHaveBeenCalled();
			select.simulate('change');
			expect(onChangeMock).toHaveBeenCalled();
		});
	});
	describe('error', () => {
		it('shows an error when error is a prop', () => {
			const props = {...MOCK_PROPS, error: 'I am erroneous' };
			const component = shallow(<FormSelect {...props} />);
			expect(component.find('.text--error').exists()).toBe(true);
			expect(component.find('.text--error').text()).toBe(props.error);
		});
	});
});
