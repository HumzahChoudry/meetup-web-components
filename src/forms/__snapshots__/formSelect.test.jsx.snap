// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelect renders correctly 1`] = `
<div>
  <label
    className="label--field required disabled labelForSonnet"
    htmlFor="sonnet"
  >
    Shakespearean Sonnet
  </label>
  <select
    className="classForSonnet"
    disabled={true}
    id="sonnet"
    name="sonnets"
    onChange={[Function]}
    required={true}
  >
    <option
      selected={false}
      value="Love is my sin and thy dear virtue hate"
    >
      Love is my sin and thy dear virtue hate
    </option>
    <option
      selected={false}
      value="Hate of my sin, grounded on sinful loving:"
    >
      Hate of my sin, grounded on sinful loving:
    </option>
    <option
      selected={false}
      value="O, but with mine compare thou thine own state,"
    >
      O, but with mine compare thou thine own state,
    </option>
    <option
      selected={false}
      value="And thou shalt find it merits not reproving;"
    >
      And thou shalt find it merits not reproving;
    </option>
    <option
      selected={false}
      value="Or, if it do, not from those lips of thine,"
    >
      Or, if it do, not from those lips of thine,
    </option>
    <option
      selected={false}
      value="That have profaned their scarlet ornaments"
    >
      That have profaned their scarlet ornaments
    </option>
    <option
      selected={false}
      value="And seal'd false bonds of love as oft as mine,"
    >
      And seal'd false bonds of love as oft as mine,
    </option>
    <option
      selected={false}
      value="Robb'd others' beds' revenues of their rents."
    >
      Robb'd others' beds' revenues of their rents.
    </option>
    <option
      selected={false}
      value="Be it lawful I love thee, as thou lovest those"
    >
      Be it lawful I love thee, as thou lovest those
    </option>
    <option
      selected={false}
      value="Whom thine eyes woo as mine importune thee:"
    >
      Whom thine eyes woo as mine importune thee:
    </option>
    <option
      selected={false}
      value="Root pity in thy heart, that when it grows"
    >
      Root pity in thy heart, that when it grows
    </option>
    <option
      selected={false}
      value="Thy pity may deserve to pitied be."
    >
      Thy pity may deserve to pitied be.
    </option>
    <option
      selected={true}
      value="If thou dost seek to have what thou dost hide,"
    >
      If thou dost seek to have what thou dost hide,
    </option>
    <option
      selected={false}
      value="By self-example mayst thou be denied!"
    >
      By self-example mayst thou be denied!
    </option>
  </select>
</div>
`;
